server/container_create_linux.go

```diff
@@ -846,25 +846,52 @@ func (s *Server) createSandboxContainer(ctx context.Context, ctr ctrfactory.Cont
 		specgen.Config.Process.User.Umask = &umask
 	}
 
-	etc := filepath.Join(mountPoint, "/etc")
-	// create the `/etc` folder only when it doesn't exist
-	if _, err := os.Stat(etc); err != nil && os.IsNotExist(err) {
+	etcPath := filepath.Join(mountPoint, "/etc")
+
+	// Warn users if the container /etc directory path points to a location
+	// that is not a regular directory. This could indicate that something
+	// might be afoot.
+	etc, err := os.Lstat(etcPath)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, err
+	}
+	if err == nil && !etc.IsDir() {
+		log.Warnf(ctx, "Detected /etc path for container %s is not a directory", ctr.ID())
+	}
+
+	// The /etc directory can be subjected to various attempts on the path (directory)
+	// traversal attacks. As such, we need to ensure that its path will be relative to
+	// the base (or root, if you wish) of the container to mitigate a container escape.
+	etcPath, err = securejoin.SecureJoin(mountPoint, "/etc")
+	if err != nil {
+		return nil, fmt.Errorf("failed to resolve container /etc directory path: %w", err)
+	}
+
+	// Create the /etc directory only when it doesn't exist.
+	if _, err := os.Stat(etcPath); err != nil && os.IsNotExist(err) {
 		rootPair := idtools.IDPair{UID: 0, GID: 0}
 		if containerIDMappings != nil {
 			rootPair = containerIDMappings.RootPair()
 		}
-		if err := idtools.MkdirAllAndChown(etc, 0o755, rootPair); err != nil {
-			return nil, fmt.Errorf("error creating mtab directory: %w", err)
+		if err := idtools.MkdirAllAndChown(etcPath, 0o755, rootPair); err != nil {
+			return nil, fmt.Errorf("failed to create container /etc directory: %w", err)
 		}
 	}
-	// add symlink /etc/mtab to /proc/mounts allow looking for mountfiles there in the container
-	// compatible with Docker
-	if err := os.Symlink("/proc/mounts", filepath.Join(etc, "mtab")); err != nil && !os.IsExist(err) {
-		return nil, err
+
+	// we cannot use SecureJoin here, as /etc/mtab is already being symlinked from somewhere in some cases
+	// and doing so resolves mtabPath to mountPoint/etc/proc/self/mounts, which breaks container creation.
+	// Let's just check if mtab is created already, and if not, symlink it.
```

Comment:
This is an interesting side effect of using `securejoin.SecureJoin()` since it will try to resolve to underlying symbolic link.

For example, our problematic container image `quay.io/crio/fedora-crio-ci:latest` already has the following:

```console
lrwxrwxrwx 1 root root         19 Aug 10  2023 mtab -> ../proc/self/mounts
```

Which cases the `SecureJoin()` function to resolve it to (an example from a failed run):

```go
(string) (len=111) "/var/lib/containers/storage/overlay/ef720b916452ac083f3c8438627d9f5ca0f8180c5142a5befcb40cb2f80045ca/merged/etc"
(string) (len=116) "/var/lib/containers/storage/overlay/ef720b916452ac083f3c8438627d9f5ca0f8180c5142a5befcb40cb2f80045ca/merged/etc/mtab"
(string) (len=128) "/var/lib/containers/storage/overlay/ef720b916452ac083f3c8438627d9f5ca0f8180c5142a5befcb40cb2f80045ca/merged/etc/proc/self/mounts"
```

However, when run under Docker (moby), the /etc/mtab symbolic link will be recreated per:

```console
lrwxrwxrwx 1 root root      12 May 29 08:11 mtab -> /proc/mounts
```

This is because of the following code:

- https://github.com/moby/moby/blob/master/daemon/initlayer/setup_unix.go#L19-L72

I wonder whether it would be safer to also remove existing and add new /etc/mtab symbolic link? The procfs nowadays creates /proc/mounts as symbolic link to /proc/self/mounts anyway (see [man 5 proc_pid_mounts](https://manpages.debian.org/unstable/manpages/proc_pid_mounts.5.en.html)), and as such, it would be safe to always create it from scratch.

Reply:
hm maybe we should remove the symlink in the image as it may actually be bogus and then readd the change?

Reply:
@haircommander that is what I am suggesting too.

Per the everpresent documentation, the `/etc/mtab`, if present, should be a symbolic link to `/proc/mounts`. There is no need for it to be different, I suppose. This is also to retain some backwards compatibility with older software and Linux distributions, plus to retain what is expected to be a canonical (not to say, historical even) thing for this specific file/symbolic link.

Reply:
okay so I did some investigation, and all the fedora family creates this symlink. It's not present everywhere or magically injected, as the ubuntu image doesn't have it. thus, I think the code as-is is correct

Reply:
Yeah, the code is fine. :+1: 

What I wanted to bring up, was that if we remove it, no matter what it is and where it's currently pointing, and add it ourselves, then we can always guarantee its correctness.

Albeit, leaving it alone is also an option.